### PR TITLE
Refactor TestCachingJdbcClient and remove unnecessary suppressions

### PR DESCRIPTION
**Pull request by wendigo**

### CODE REVIEW

File: plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java [hunk 1]
```diff
@@ -176,18 +174,18 @@ public void testColumnsCached()
         JdbcTableHandle table = getAnyTable(schema);
         JdbcColumnHandle phantomColumn = addColumn(table);
 
-        int expectedLoad = 0, expectedHit = 0, expectedMiss = 0;
+        AssertedStats stats = new AssertedStats();
 
         // Read column into cache
         assertThat(cachingJdbcClient.getColumns(SESSION, table)).contains(phantomColumn);
-        assertCacheLoadsHitsAndMisses(cachingJdbcClient.getColumnsCacheStats(), expectedLoad += 1, expectedHit, expectedMiss += 1);
+        assertCacheLoadsHitsAndMisses(cachingJdbcClient.getColumnsCacheStats(), stats.addLoads(1).addMisses(1));
```
Comment (ssheikin) [inline plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java:181]: Why do we want both of this methods: `addLoads()` and `addMisses()`?

**Review by losipiuk — APPROVED** [review]

File: plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java (hunk 1 above)
Comment (wendigo) [reply plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java:181]: What do you propose instead? They are increasing different counters. This change does not decrease code readability

File: plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java [hunk 2]
```diff
@@ -198,19 +196,19 @@ public void testColumnsCachedPerSession()
         JdbcTableHandle table = getAnyTable(schema);
         JdbcColumnHandle phantomColumn = addColumn(table);
 
-        int expectedLoad = 0, expectedHit = 0, expectedMiss = 0;
+        AssertedStats stats = new AssertedStats();
```
Comment (kokosing) [inline plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java:199]: That should get `cachingJdbcClient.getColumnsCacheStats()` as parameter.

```
assertStats = asserStats(cachingJdbcClient.getColumnsCacheStats());
assertStats.addLoaded(1).addMisses(1);
```

Each `add*` could do already assertion..

**Comment by wendigo** [comment]
Woah, those failures...

```
Current runner version: '2.276.1'
Operating System
Virtual Environment
Prepare workflow directory
Prepare all required actions
Getting action download info
Download action repository 'actions/checkout@v2'
Download action repository 'actions/setup-java@v1'
Error: An action could not be found at the URI 'https://api.github.com/repos/actions/setup-java/tarball/d202f5dbf7256730fb690ec59f6381650114feb2'
```

File: plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java (hunk 1 above)
Comment (ssheikin) [reply plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java:181]: Just asking

File: plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java (hunk 2 above)
Comment (wendigo) [reply plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java]: PTAL now. I've made a fixup that I will squash into final form if we agree on solution.

File: plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java [hunk 3]
```diff
@@ -196,19 +199,21 @@ public void testColumnsCachedPerSession()
         JdbcTableHandle table = getAnyTable(schema);
         JdbcColumnHandle phantomColumn = addColumn(table);
 
-        AssertedStats stats = new AssertedStats();
+        CacheStatsAssertions statsAssertions = new CacheStatsAssertions(cachingJdbcClient.getColumnsCacheStats());
 
         // Load columns in first session scope
         assertThat(cachingJdbcClient.getColumns(firstSession, table)).contains(phantomColumn);
-        assertCacheLoadsHitsAndMisses(cachingJdbcClient.getColumnsCacheStats(), stats.addLoads(1).addMisses(1));
+        statsAssertions.loadsIncreasedBy(1);
```
Comment (ssheikin) [inline plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java:206]: Chaining here was nice :(

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/CachingJdbcClient.java [hunk 4]
```diff
@@ -412,9 +413,9 @@ private void invalidateColumnsCache(SchemaTableName table)
     }
 
     @VisibleForTesting
-    CacheStats getColumnsCacheStats()
+    Supplier<CacheStats> getColumnsCacheStats()
```
Comment (ssheikin) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/CachingJdbcClient.java:416]: may be don't touch this class and make supplier when invoked?

File: plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java [hunk 5]
```diff
@@ -369,43 +375,38 @@ public void testEverythingImplemented()
...
-
-        public int getMisses()
+        public CacheStatsAssertions(Supplier<CacheStats> stats)
         {
-            return misses;
+            this.stats = requireNonNull(stats, "stats is null");
         }
 
-        public AssertedStats addLoads(int value)
+        public void loadsIncreasedBy(int value)
```
Comment (kokosing) [inline plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java:391]: return `CacheStatsAssertions` so we can chain

File: plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java [hunk 6]
```diff
@@ -369,43 +375,38 @@ public void testEverythingImplemented()
...
+            assertThat(stats.get().loadCount()).withFailMessage("Expected load count is %d but actual is %d", loads, stats.get().loadCount())
+                    .isEqualTo(loads);
         }
 
-        public AssertedStats addHits(int value)
+        public void hitsIncreasedBy(int value)
         {
             this.hits += value;
-            return this;
+            assertThat(stats.get().hitCount()).withFailMessage("Expected hit count is %d but actual is %d", hits, stats.get().hitCount())
```
Comment (kokosing) [inline plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java:401]: put `.withFailMessage` to the new line?

File: plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java [hunk 7]
```diff
@@ -283,19 +293,23 @@ public void testColumnsNotCachedWhenCacheDisabled()
...
+        cacheAssertions.missesIncreasedBy(4);
 
         assertThat(cachingJdbcClient.getColumns(firstSession, firstTable)).containsExactly(firstColumn);
         assertThat(cachingJdbcClient.getColumns(secondSession, firstTable)).containsExactly(firstColumn);
         assertThat(cachingJdbcClient.getColumns(firstSession, secondTable)).containsExactly(secondColumn);
         assertThat(cachingJdbcClient.getColumns(secondSession, secondTable)).containsExactly(secondColumn);
-        assertCacheLoadsHitsAndMisses(cachingJdbcClient.getColumnsCacheStats(), stats.addLoads(4).addMisses(4));
+
+        cacheAssertions.loadsIncreasedBy(4);
+        cacheAssertions.missesIncreasedBy(4);
```
Comment (kokosing) [inline plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java:312]: Let's chain them.

**Review by kokosing — COMMENTED** [review]
I like it. Thanks!

File: plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java (hunk 6 above)
Comment (wendigo) [reply plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java:401]: Sure.

File: plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java [hunk 8]
```diff
@@ -285,19 +285,21 @@ public void testColumnsNotCachedWhenCacheDisabled()
...
+        cacheAssertions.loadsIncreasedBy(4).missesIncreasedBy(4);
 
         assertThat(cachingJdbcClient.getColumns(firstSession, firstTable)).containsExactly(firstColumn);
         assertThat(cachingJdbcClient.getColumns(secondSession, firstTable)).containsExactly(firstColumn);
         assertThat(cachingJdbcClient.getColumns(firstSession, secondTable)).containsExactly(secondColumn);
         assertThat(cachingJdbcClient.getColumns(secondSession, secondTable)).containsExactly(secondColumn);
-        assertCacheLoadsHitsAndMisses(cachingJdbcClient.getColumnsCacheStats(), expectedLoad += 4, expectedHit, expectedMiss += 4);
+        cacheAssertions
+                .loadsIncreasedBy(4)
+                .missesIncreasedBy(4);
```
Comment (ssheikin) [inline plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java:302]: join lines, as above.

File: plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java [hunk 9]
```diff
@@ -306,17 +308,9 @@ public void testColumnsNotCachedWhenCacheDisabled()
...
-    {
-        assertThat(stats.loadCount()).withFailMessage("Expected load count is %d but actual is %d", expectedLoad, stats.loadCount())
-                .isEqualTo(expectedLoad);
-        assertThat(stats.hitCount()).withFailMessage("Expected hit count is %d but actual is %d", expectedHit, stats.hitCount())
-                .isEqualTo(expectedHit);
-        assertThat(stats.missCount()).withFailMessage("Expected miss count is %d but actual is %d", expectedMiss, stats.missCount())
-                .isEqualTo(expectedMiss);
+        cacheAssertions
+                .loadsIncreasedBy(2)
+                .missesIncreasedBy(2);
```
Comment (ssheikin) [inline plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java:313]: join lines, as above.

**Review by ssheikin — APPROVED** [review]

File: plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java (hunk 8 above)
Comment (wendigo) [reply plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java:302]: It's more readable this way

**Comment by kokosing** [comment]
Automation hit: https://github.com/trinodb/trino/issues/6799

File: plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java [hunk 10]
```diff
@@ -382,4 +360,45 @@ public void testEverythingImplemented()
...
+                    .isEqualTo(hits);
+            return this;
+        }
+
+        public CacheStatsAssertions missesIncreasedBy(int value)
+        {
+            this.misses += value;
+            assertThat(stats.get().missCount())
+                    .withFailMessage("Expected miss count is %d but actual is %d", misses, stats.get().missCount())
+                    .isEqualTo(misses);
```
Comment (kokosing) [inline plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java:400]: We should also verify other stats. If you verify that `missesIncreasedBy` then you need to verify that load and hits didn't change. 

Maybe you need `loadsAndMissesIncreasedBy(2)`?

File: plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java [hunk 11]
```diff
@@ -79,7 +79,7 @@ public void setUp()
 
     private CachingJdbcClient createCachingJdbcClient(Duration cacheTtl, boolean cacheMissing)
     {
-        return new CachingJdbcClient(database.getJdbcClient(), Set.of(getTestSessionPropertiesProvider()), cacheTtl, cacheMissing);
+        return new CachingJdbcClient(database.getJdbcClient(), Set.of(() -> PROPERTY_METADATA), cacheTtl, cacheMissing);
```
Comment (kokosing) [inline plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java:82]: Please extract private constant for ` Set.of(() -> PROPERTY_METADATA)`, PROPERTY_METADATA_PROVIDERS?

Comment (wendigo) [reply plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java]: Done

File: plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java (hunk 10 above)
Comment (wendigo) [reply plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java:405]: PTAL now

Comment (ssheikin) [reply plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java:405]: @kokosing, I think that it is just the duplication of logic. I'd get rid of that `verify()`

**Review by ssheikin — APPROVED** [review]

File: plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java (hunk 10 above)
Comment (kokosing) [reply plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java:405]: We always need to verify all three metrics (or two, to me misses=loaded), however I don't like this additional call.

> (or two, to me misses=loaded)

If we would remove `loaded` metric verification, then you can inline `verify()` into `missesIncreasedBy()` and `hitsIncreasedBy()`

Comment (wendigo) [reply plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java:405]: I'm fine with a little duplicatiob

File: plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java [hunk 12]
```diff
@@ -367,57 +385,57 @@ public void testEverythingImplemented()
...
+            CacheStats afterStats = stats.get();
+
+            long expectedLoad = beforeStats.loadCount() + loads;
+            long expectedMisses = beforeStats.missCount() + misses;
+            long expectedHits = beforeStats.hitCount() + hits;
+
+            // Reset expectations
+            loads = 0;
+            misses = 0;
+            hits = 0;
```
Comment (kokosing) [inline plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java:428]: [optional] I don't like this statefullness here. Maybe it should not be reused?

**Review by kokosing — APPROVED** [review]
I added a comment, but I am ok with merging this as it is, as it is nice improvement anyway.

File: plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java [hunk 13]
```diff
@@ -181,15 +181,18 @@ public void testColumnsCached()
         CacheStatsAssertions statsAssertions = new CacheStatsAssertions(cachingJdbcClient::getColumnsCacheStats);
```
[the remark this answers is not in the document]
Comment (wendigo) [reply plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java:181]: Fixed. Sorry.

**Review by kokosing — APPROVED** [review]

**Comment by kokosing** [comment]
Merged as: https://github.com/trinodb/trino/commit/5a3b42879c71c72cff5bbfd540554f190654ae7a

**Comment by kokosing** [comment]
Thanks!

**Pull request closed without merging**